### PR TITLE
Feature/file migration

### PR DESCRIPTION
**Pull request by SeniorZhai**

### CODE REVIEW

File: app/src/main/java/one/mixin/android/db/MessageDao.kt [hunk 1]
```diff
@@ -495,10 +495,16 @@ interface MessageDao : BaseDao<Message> {
     suspend fun batchQueryMessages(limit: Int, offset: Int, after: Long): List<QueryMessage>
 
     @Query(
-        "SELECT id, conversation_id, name, category, media_url, media_mine_type FROM messages WHERE category IN ('SIGNAL_IMAGE','PLAIN_IMAGE', 'SIGNAL_VIDEO', 'PLAIN_VIDEO', 'SIGNAL_DATA', 'PLAIN_DATA', 'SIGNAL_AUDIO', 'PLAIN_AUDIO') AND media_status = 'DONE' AND rowid <= :rowId LIMIT :limit OFFSET :offset"
+        "SELECT id, conversation_id, name, category, media_url, media_mine_type FROM messages WHERE category IN ('SIGNAL_IMAGE','PLAIN_IMAGE', 'SIGNAL_VIDEO', 'PLAIN_VIDEO', 'SIGNAL_DATA', 'PLAIN_DATA', 'SIGNAL_AUDIO', 'PLAIN_AUDIO', 'PLAIN_TRANSCRIPT', 'SIGNAL_TRANSCRIPT') AND media_status = 'DONE' AND rowid <= :rowId LIMIT :limit OFFSET :offset"
     )
     fun findAttachmentMigration(rowId: Long, limit: Int, offset: Long): List<AttachmentMigration>
 
+    @Query("SELECT count(1) FROM messages WHERE category IN ('SIGNAL_IMAGE','PLAIN_IMAGE', 'SIGNAL_VIDEO', 'PLAIN_VIDEO', 'SIGNAL_DATA', 'PLAIN_DATA', 'SIGNAL_AUDIO', 'PLAIN_AUDIO', 'PLAIN_TRANSPORT') AND media_status = 'DONE'")
```
Comment (Tougee) [inline app/src/main/java/one/mixin/android/db/MessageDao.kt]: typo PLAIN_TRANSPORT?
missing SIGNAL_TRANSCRIPT?

File: app/src/main/java/one/mixin/android/db/MessageDao.kt [hunk 2]
```diff
@@ -495,10 +495,16 @@ interface MessageDao : BaseDao<Message> {
...
+    fun countDoneAttachment(): Int
+
+    @Query("SELECT rowid IS NOT NULL FROM transcript_messages WHERE category IN ('SIGNAL_IMAGE','PLAIN_IMAGE', 'SIGNAL_VIDEO', 'PLAIN_VIDEO', 'SIGNAL_DATA', 'PLAIN_DATA', 'SIGNAL_AUDIO', 'PLAIN_AUDIO', 'PLAIN_TRANSPORT') AND media_status = 'DONE' LIMIT 1")
```
Comment (Tougee) [inline app/src/main/java/one/mixin/android/db/MessageDao.kt]: as above

File: app/src/main/java/one/mixin/android/extension/FileExtension.kt [hunk 3]
```diff
@@ -68,32 +69,29 @@ fun Context.checkStorageNotLow(lowAction: () -> Unit, defaultAction: () -> Unit)
...
-fun Context.getMediaPath(): File? {
-    val path = getAppPath() ?: return null
+fun Context.getMediaPath(legacy: Boolean = false): File? {
+    val path = getAppPath(legacy) ?: return null
     val identityNumber = Session.getAccount()?.identityNumber ?: return null
     return File("${path.absolutePath}${File.separator}$identityNumber${File.separator}Media")
 }
 
-fun Context.getOldMediaPath(): File? {
+fun Context.getAncientMediaPath(): File? {
```
Comment (Tougee) [inline app/src/main/java/one/mixin/android/extension/FileExtension.kt:94]: LOL

File: app/src/main/java/one/mixin/android/job/MixinJobManager.kt [hunk 4]
```diff
@@ -10,7 +10,7 @@ import java.util.concurrent.ConcurrentHashMap
 class MixinJobManager(configuration: Configuration) : JobManager(configuration) {
 
     companion object {
-        val attachmentProcess = ArrayMap<String, Int>()
+        var attachmentProcess = ArrayMap<String, Int>()
```
Comment (Tougee) [inline app/src/main/java/one/mixin/android/job/MixinJobManager.kt]: seems an unnecessary change

File: app/src/main/java/one/mixin/android/job/TranscriptAttachmentDownloadJob.kt [hunk 5]
```diff
@@ -197,7 +196,7 @@ class TranscriptAttachmentDownloadJob(
                     }
                     imageFile.copyFromInputStream(attachmentCipherInputStream)
                     transcriptMessageDao.updateMedia(
-                        Uri.fromFile(imageFile).toString(),
+                        imageFile.name,
```
Comment (Tougee) [inline app/src/main/java/one/mixin/android/job/TranscriptAttachmentDownloadJob.kt:199]: use file name as media url?

File: app/src/main/java/one/mixin/android/extension/FileExtension.kt [hunk 6]
```diff
@@ -102,8 +100,8 @@ fun Context.getOldMediaPath(): File? {
     return null
 }
 
-fun Context.getBackupPath(create: Boolean = false): File? {
-    val path = getAppPath() ?: return null
+fun Context.getLegacyBackupPath(create: Boolean = false, legacy: Boolean = false): File? {
```
Comment (Tougee) [inline app/src/main/java/one/mixin/android/extension/FileExtension.kt:103]: keep name as `getBackupPath`?

File: app/src/main/java/one/mixin/android/job/BackupMigrationJob.kt [hunk 7]
```diff
@@ -1,35 +1,12 @@
...
-import one.mixin.android.MixinApplication
-import one.mixin.android.extension.defaultSharedPreferences
-import one.mixin.android.extension.getBackupPath
-import one.mixin.android.extension.hasWritePermission
-import one.mixin.android.extension.putBoolean
-import one.mixin.android.util.backup.findOldBackupSync
-import timber.log.Timber
-import java.io.File
 
 class BackupMigrationJob : BaseJob(Params(PRIORITY_LOWER).groupBy(GROUP_ID).persist()) {
```
Comment (Tougee) [inline app/src/main/java/one/mixin/android/job/BackupMigrationJob.kt:5]: can this class be removed?

File: app/src/main/java/one/mixin/android/util/MediaStoreUtil.kt [hunk 8]
```diff
@@ -0,0 +1,228 @@
...
+import one.mixin.android.extension.getOtherPath
+import one.mixin.android.extension.getVideoPath
+import one.mixin.android.extension.isImageSupport
+import one.mixin.android.session.Session
+import one.mixin.android.widget.gallery.MimeType
+import java.io.File
+import java.io.IOException
+import java.io.InputStream
+
+fun getBackupPath(): File? {
```
Comment (Tougee) [inline app/src/main/java/one/mixin/android/util/MediaStoreUtil.kt]: this file seems not in use

File: app/src/main/java/one/mixin/android/util/PropertyHelper.kt [hunk 9]
```diff
@@ -0,0 +1,126 @@
...
+import one.mixin.android.Constants.Download.ROAMING_DEFAULT
+import one.mixin.android.Constants.Download.WIFI_DEFAULT
+import one.mixin.android.db.MessageDao
+import one.mixin.android.db.MixinDatabase
+import one.mixin.android.db.PropertyDao
+import one.mixin.android.extension.defaultSharedPreferences
+import one.mixin.android.extension.nowInUtc
+import one.mixin.android.vo.Property
+
+object PropertyHelper {
```
Comment (Tougee) [inline app/src/main/java/one/mixin/android/util/PropertyHelper.kt:30]: use MixinApplication.appContext instead of a pass-in context  seems better to prevent Activity context leaks

File: app/src/main/java/one/mixin/android/vo/ChatHistoryMessageItem.kt [hunk 10]
```diff
@@ -187,12 +183,11 @@ fun ChatHistoryMessageItem.absolutePath(context: Context = MixinApplication.appC
...
         }
     }
-    val mediaPath = MixinApplication.appContext.getMediaPath()?.toUri()?.toString() ?: return null
     val url = mediaUrl
     return when {
-        url == null -> null
+        url == null || mediaUrl == null -> null
         isLive() -> url
-        url.startsWith(mediaPath) -> url
+        mediaUrl?.isFileUri() == true -> url
```
Comment (Tougee) [inline app/src/main/java/one/mixin/android/vo/ChatHistoryMessageItem.kt]: mediaUrl.isFileUri() -> url

File: app/src/main/java/one/mixin/android/widget/gallery/internal/loader/AlbumLoader.java [hunk 11]
```diff
@@ -93,7 +93,7 @@
     // =============================================
 
     // TODO workaround, need refactor loading
-    private static final String BUCKET_ORDER_BY = "datetaken DESC LIMIT 5000";
```
Comment (Tougee) [inline app/src/main/java/one/mixin/android/widget/gallery/internal/loader/AlbumLoader.java:96]: keep this for workaround?

File: app/src/main/java/one/mixin/android/ui/home/MainActivity.kt [hunk 12]
```diff
@@ -266,53 +258,81 @@ class MainActivity : BlazeBaseActivity() {
...
+            jobManager.addJobInBackground(RefreshFiatsJob())
+        }
 
-            if (Fiats.isRateEmpty()) {
-                jobManager.addJobInBackground(RefreshFiatsJob())
-            }
+        PropertyHelper.checkFts4Upgrade(this@MainActivity) {
+            InitializeActivity.showFts(this@MainActivity)
+            finish()
+        }
```
Comment (Tougee) [inline app/src/main/java/one/mixin/android/ui/home/MainActivity.kt]: move to the top of this method and add a `return` could be better?

File: app/src/main/java/one/mixin/android/widget/gallery/internal/loader/AlbumLoader.java (hunk 11 above)
Comment (SeniorZhai) [reply app/src/main/java/one/mixin/android/widget/gallery/internal/loader/AlbumLoader.java:96]: There should be a change here, it caused a crash on my device

File: app/src/main/java/one/mixin/android/job/TranscriptAttachmentDownloadJob.kt (hunk 5 above)
Comment (SeniorZhai) [reply app/src/main/java/one/mixin/android/job/TranscriptAttachmentDownloadJob.kt:199]: use absolute path
